### PR TITLE
Optimization density_on_grid_update_impl round 1

### PR DESCRIPTION
**Pull request by bE554357**
Код с кое-какой валидацией. Время на 1 потоке:
```
reference: 2[calls], 5.00747[s], 2.50373[s/call]
optimized: 2[calls], 0.813149[s], 0.406574[s/call]
```
на 4 потоках:
```
reference: 2[calls], 1.35147[s], 0.675736[s/call]
optimized: 2[calls], 0.207251[s], 0.103625[s/call]
```
на 8 потоках
```
reference: 2[calls], 1.66664[s], 0.833322[s/call]
optimized: 2[calls], 0.109725[s], 0.0548627[s/call]
```

Тот код, который сейчас в мастере, по сути не параллельный. А мои оптимизации уверенно хороши только для малого числа потоков, т.к. есть сомнения в скорости работы на большом количестве ядер и для MPI ничего не настроено. Также есть вопросы к выделению памяти, но в любом случае вижу такой алгоритм в этом месте:

1. Каждый поток считает данные локально на всём поле или на одном его фрагменте (разбить частицы по областям линейная требует линейное время от количества частиц).
2. Делается allreduce: самописный внутри процесса или `MPI_Allreduce` между процессами.
3. И так по всем областям.

И никаких атомарных операций, кроме выставления флагов для синхронизации при межпотоковом reduce.

Перед заливкой надо только удалить валидацию.

### CODE REVIEW

**Comment by bE554357** [comment]
@eaberendeev

**Comment by bE554357** [comment]
V2 --- моя изначальная оптимизаций.

V4 --- частицы собираются по потокам:

V7 это оптимизированная V4.


Времена на 1 вызов (вызова 2 на шаг) в текущём мастере:

| nthr | Ref | V2 | V4 | V7 |
|--------|--------|--------|--------|--------|
| 1 | 0.312 | 0.051 | 0.110 |0.047|
| 2 | 0.169 | 0.025 | 0.056 |0.024|
| 4 | 0.092 | 0.013| 0.028 |0.013|
| 8 | 0.079 |0.007 |0.015 | 0.007|

**Comment by bE554357** [comment]
Время на кластере:
```
###############################
move_and_calc_current_impl
###############################
SMAX: 4
ref error: 1.0258e-16, normalized 2.56304e-17
opt V8 error: 5.02265e-15, normalized 1.25495e-15
opt V9A error: 1.2079e-16, normalized 3.01804e-17
opt V9B error: 1.17046e-16, normalized 2.92448e-17
opt V9C error: 5.76432e-16, normalized 1.44026e-16
opt V10 error: 5.76488e-16, normalized 1.4404e-16
opt V11 error: 5.76868e-16, normalized 1.44135e-16
opt V12 error: 5.02936e-15, normalized 1.25663e-15
Applying boundary conditions to Electrons
Added 57 Electrons particles from emitter
Time to apply boundary conditions: 5.41815 seconds
Current 2.00114
Time: 6.13781
Time1: 2.29048
Correction fieldE solver error = 4.77473e-10
Correction fieldE Mysolver time = 6.3482
lambda 1.00001 1.00001
Ref vs test V2, norm: 4.33175e-12, normalized: 3.44559e-15
Ref vs test V4, norm: 4.27282e-12, normalized: 3.39872e-15
Ref vs test V7, norm: 4.3688e-12, normalized: 3.47507e-15
1257.19 norm mesh.chargeDensity 
6.41571e-13 norm drho / Dt - divJ 
Moved 965097263 Electrons
Collision time: 2.37022e-06
Energy 0.235656 -0.235656 0.235656 0
151.5 4710.59 4710.83 9.65097e+08 0 0 0 0 0 1568.9 3141.93 10.2319 0.615625 163493 2.26766e-10 
> all: 1[calls], 867.404[s]
|  > sim loop: 1[calls], 867.404[s]
|  |  > Prepare step: 1[calls], 854.168[s]
|  |  |  > virtual void SimulationEcsim::prepare_step: 1[calls], 1.39985[s]
|  |  |  > virtual void SimulationEcsimCorr::make_step: 1[calls], 852.768[s]
|  |  |  |  > void ParticlesArray::density_on_grid_update_impl: 2[calls], 29.5391[s], 14.7696[s/call]
|  |  |  |  |  > reference: 2[calls], 19.7592[s], 9.87959[s/call]
|  |  |  |  |  > test V4: 2[calls], 4.38265[s], 2.19132[s/call]
|  |  |  |  |  > test V7: 2[calls], 2.15874[s], 1.07937[s/call]
|  |  |  |  |  > test V2: 2[calls], 3.10658[s], 1.55329[s/call]
|  |  |  |  \  > self : 0.131952[s]
|  |  |  |  > void ParticlesArray::move_and_calc_current_impl: 2[calls], 752.568[s], 376.284[s/call]
|  |  |  |  |  > reference code: 2[calls], 10.5865[s], 5.29327[s/call]
|  |  |  |  |  > optimized V8: 2[calls], 12.2781[s], 6.13904[s/call]
|  |  |  |  |  > optimized V9A: 2[calls], 8.9073[s], 4.45365[s/call]
|  |  |  |  |  > optimized V9B: 2[calls], 9.05371[s], 4.52685[s/call]
|  |  |  |  |  > optimized V9C: 2[calls], 7.54436[s], 3.77218[s/call]
|  |  |  |  |  > optimized V10: 2[calls], 5.85878[s], 2.92939[s/call]
|  |  |  |  |  > optimized V11: 2[calls], 5.77226[s], 2.88613[s/call]
|  |  |  |  |  > optimized V12: 2[calls], 7.62955[s], 3.81478[s/call]
|  |  |  |  |  > original ref code: 2[calls], 10.636[s], 5.31801[s/call]
|  |  |  |  \  > self : 674.301[s]
|  |  |  |  > void algorithmsECSIM::predict_current_impl_linear: 1[calls], 1.37878[s]
|  |  |  |  > void ParticlesArray::fill_matrixL_impl_linear2: 1[calls], 6.18838[s]
|  |  |  |  > void Mesh::stencil_Lmat2: 1[calls], 12.3704[s]
|  |  |  |  |  > parallel sections: 1[calls], 5.3385[s]
|  |  |  |  |  > section 2: 1[calls], 5.52676[s]
|  |  |  |  |  > section 3: 1[calls], 5.281e-06[s]
|  |  |  |  |  > void optimizedSetFromTriplets: 1[calls], 1.4216[s]
|  |  |  |  \  > self : 0.0834929[s]
|  |  |  |  > virtual void OpenBoundaryCondition::apply_to_operator: 1[calls], 0.312656[s]
|  |  |  |  > void SimulationEcsim::predict_electric_field: 1[calls], 20.11[s]
|  |  |  |  |  > bool bicgstab_iteration: 1[calls], 11.3901[s]
|  |  |  |  |  |  > void spmv: 87[calls], 7.69773[s], 0.0884796[s/call]
|  |  |  |  |  \  > self : 3.69234[s]
|  |  |  |  \  > self : 8.71998[s]
|  |  |  |  > bool bicgstab_iteration: 1[calls], 6.29187[s]
|  |  |  |  |  > void spmv: 89[calls], 2.32108[s], 0.0260795[s/call]
|  |  |  |  \  > self : 3.97079[s]
|  |  |  \  > self : 24.0093[s]
|  |  \  > self : 1.4061e-05[s]
|  |  > loop 1: 1[calls], 0.0094477[s]
|  |  > collision step: 1[calls], 5.16e-06[s]
|  |  > diagnostic: 1[calls], 13.2263[s]
|  |  |  > virtual void SimulationEcsim::make_diagnostic: 1[calls], 13.2196[s]
|  |  \  > self : 0.00673943[s]
|  \  > self : 1.1331e-05[s]
\  > self : 1.817e-05[s]
```

**Comment by eaberendeev** [comment]
сделай flat таймером табличку для теста, чтобы сразу были видны времена

**Comment by bE554357** [comment]
> сделай flat таймером табличку для теста, чтобы сразу были видны времена

Теперь пишется такое в конца каждого теста:
image

File: srcBeren/algorithms/decompose_esirkepov_current.h [hunk 1]
```diff
@@ -473,3 +473,129 @@ inline void move_and_calc_current_reference(Array3D<std::vector<Particle>>& part
...
+#pragma omp atomic update
+                        fieldJ(indx, indy, indz, 2) += jz[n][m][k];
+                    }
+                }
+            }
+        }
+    }
+}
+
+inline void move_and_calc_current_original(Array3D<std::vector<Particle>>& particlesData, const Domain& domain,
```
Comment (bE554357) [inline srcBeren/algorithms/decompose_esirkepov_current.h:586]: Зачем тут это, когда есть `move_and_calc_current_original`? @eaberendeev

Comment (eaberendeev) [reply srcBeren/algorithms/decompose_esirkepov_current.h:586]: Это оригинальный алгоритм, можно его хранить как референс. Не знаю короче.

Comment (eaberendeev) [reply srcBeren/algorithms/decompose_esirkepov_current.h:586]: Вынести в tests референсные версии функций для тестирования
